src/Interpreters/Cache/LRUFileCachePriority.cpp

```diff
@@ -110,19 +137,31 @@ void LRUFileCachePriority::iterate(IterateFunc && func)
     }
 }
 
-LRUFileCachePriority::Iterator LRUFileCachePriority::LRUFileCacheIterator::remove()
+LRUFileCachePriority::Iterator LRUFileCachePriority::LRUFileCacheIterator::remove(const CacheGuard::Lock &)
 {
     return std::make_shared<LRUFileCacheIterator>(cache_priority, cache_priority->remove(queue_iter));
 }
 
-void LRUFileCachePriority::LRUFileCacheIterator::incrementSize(ssize_t size)
+void LRUFileCachePriority::LRUFileCacheIterator::annul()
```

Comment:
Found this method very strange. We decrement priority queue size here, so we do when erasing the cache entry from cache. So, the `current_size` will end up incorrect.
What the purpose of this method at all? Logically, we should either detach the segment or remove it from the cache. But this method does something in the middle.

Reply:
I guess I've got the idea, - entries with zero size are supposed to be removed within next `iterate()` [here](https://github.com/ClickHouse/ClickHouse/pull/44985/files#diff-032e65df4e63e386795ac9c91c92ec6a611cc6493429a9343a720ed224d93f60R98).

The issue is that access to `it->size` is not synchronized, so basically it's a data race. Let's make atomic, and add the comment to the `annul()` method and place where entries with zero size are removed during iteration